src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableOutputConnection.cs

```diff
@@ -0,0 +1,362 @@
+                if ((transferred > 0) && Strategy.ProcessTransferred(transferred, quotaRemaining))
+                {
+                    ActionItem.Schedule(s_sendRetries, this);
+                }
+                else
+                {
+                    OnTransferComplete();
+                }
+            }
+            else
+            {
+                WsrmFault fault = new InvalidAcknowledgementFault(_id, ranges);
+                RaiseFault(fault.CreateException(), fault);
+            }
+        }
+
+        public void ProcessTransferred(SequenceRangeCollection ranges, int quotaRemaining)
+        {
+            bool invalidAck;
+            bool inconsistentAck;
+
+            Strategy.ProcessAcknowledgement(ranges, out invalidAck, out inconsistentAck);
+
+            if (!invalidAck && !inconsistentAck)
+            {
+                if (Strategy.ProcessTransferred(ranges, quotaRemaining))
+                {
+                    ActionItem.Schedule(s_sendRetries, this);
+                }
+                else
+                {
+                    OnTransferComplete();
+                }
+            }
+            else
+            {
+                WsrmFault fault = new InvalidAcknowledgementFault(_id, ranges);
+                RaiseFault(fault.CreateException(), fault);
+            }
+        }
+
+        private void RaiseFault(Exception faultException, WsrmFault fault)
+        {
+            ComponentFaultedHandler handler = Faulted;
+
+            if (handler != null)
+                handler(faultException, fault);
+        }
+
+        private void RaiseOnException(Exception exception)
+        {
+            ComponentExceptionHandler handler = OnException;
+
+            if (handler != null)
+                handler(exception);
+        }
+
+        private async void SendRetries()
+        {
+            while (true)
+            {
+                if (_sendGuard.Enter())
+                {
+                    try
+                    {
+                        MessageAttemptInfo attemptInfo = Strategy.GetMessageInfoForRetry(false);
+                        if (attemptInfo.Message == null)
+                        {
+                            break;
+                        }
+
+                        await _sendAsyncHandler(attemptInfo, _sendTimeout, true);
+                    }
+                    finally
+                    {
+                        _sendGuard.Exit();
+                    }
+
+                    Strategy.DequeuePending();
+                    OnTransferComplete();
+                }
+                else
+                {
+                    return;
+                }
+            }
+        }
+
+        private static void SendRetries(object state)
+        {
+            ReliableOutputConnection outputConnection = (ReliableOutputConnection)state;
+
+            try
+            {
+                outputConnection.SendRetries();
+            }
+            catch (Exception e)
```

Comment:
This try/catch is *not* going to catch exceptions that happen once the async SendRetries method goes async. I'm guessing you want to move the try/catch to the asycn SendRetries method.

Reply:
You're right. This was too literal a translation of the Begin code path of the APM code as that had another try/catch on the callback method which I missed.